### ISSUE TITLE
Infer V of @final TypedDict as Union[*field_annotations] instead of object

### ISSUE THREAD

**Issue by MarioIshac** [microsoft/pyright#3297]
**Describe the bug**
Not sure if this is a feature request or bug report but given
```
from collections.abc import Mapping
from typing import TypedDict, final

@final
class MyDict(TypedDict):
    a: int

def consume_type(a: Mapping[str, int]):
    pass

m: MyDict = {"a": 1}
consume_type(m)
```
The resulting error from **Pyright 1.1.235** is:
```
error: Argument of type "MyDict" cannot be assigned to parameter "a" of type "Mapping[str, int]" in function "consume_type"
    TypeVar "_VT_co@Mapping" is covariant
      "object" is incompatible with "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```
**Expected behavior**
`MyDict` should be inferred as a subclass of `Mapping[str, int]` instead of `Mapping[str, object]`. While the current behavior is allowed per [Type Consistency ](https://peps.python.org/pep-0589/#type-consistency) of PEP 0589, having the stricter narrowing for `@final` `TypedDict`s in my eyes still satisfies the PEP. 

**Additional context**
I believe this issue is separate from #1899 because that concerns `Union` of `TypedDict`s rather than making a particular `TypedDict` also a `Mapping[str, Union[*field_annotations]]`. #3226 is a broader version of this issue which understandably was rejected because allowing this issue's requested inference for non-`@final` `TypedDict`s is not type safe.

My usecase for this feature is allowing `TypedDict`s to be considered serializable if all their instance fields are serializable themselves. For example, if the above example snippet were to be allowed, so would below:

```
from collections.abc import Mapping, Sequence
from typing import TypedDict, final

SerializableToJSON = Mapping[str, "SerializableToJSON"] | Sequence["SerializableToJSON"] | str | int | float | bool | None

@final
class MyDict(TypedDict):
    a: int

m: MyDict = {"a": 1}

def serialize(serializable: SerializableToJSON):
    pass

serialize(m)
```

**Comment by erictraut** [comment]
Pyright isn't inferring the type. It uses the `_TypedDict` class declared in `typing.pyi`. This class indicates that `TypedDict` derives from `Mapping[str, object]`. Other Python type checkers like mypy use the same technique. 

It would be theoretically possible to add special-case logic to override the `_TypedDict` class hierarchy such that second type argument to the `Mapping` base class is derived from the union of all fields defined in the `TypedDict`, but I don't think that would generally be very useful because most TypedDicts have a mixture of value types, and a union of those types isn't generally what one would expect.

So I think it's best to leave the behavior as is.

As a workaround, you can write a custom conversion routine for your own TypedDicts if they have homogeneous field types.

**Comment by MarioIshac** [comment]
It makes sense that the con of special-case logic outweighs the benefit in the admittedly rare usecases like the `SerializableToJSON` example. Thanks for your quick response here.

> As a workaround, you can write a custom conversion routine for your own TypedDicts if they have homogeneous field types.

Can you expand on what this custom conversion routine would look like? Some of my `TypedDict`s do have homogeneous field types.

**Comment by erictraut** [comment]
I'm thinking of something simple like this:
```python
def mydict_to_mapping(d: MyDict) -> Mapping[str, int]:
    return cast(dict[str, int], d)
```
